src/manager.S

```diff
@@ -129,3 +129,67 @@ add_compartment:
     b.gt      abort
 
     ret
+
+/**
+ * Function to delete an existing compartment data
+ *
+ * @param x0 Function of compartment to be deleted
+ */
+/* TODO there is a potential issue of a compartment saving a capability it was
+ * provided to it for use (i.e., PCC, DDC), such as by storing it in the memory
+ * of another compartment. If the original compartment gets deleted, and a new
+ * compartment is allocated in the same memory region, we have leaked a
+ * capability referencing the new compartment. */
+.type del_compartment, "function"
+del_compartment:
+    // Find capabilities (DDC/PCC) of given compartment
+    ldr       x1, comps_addr
+    mov       w2, #0
+    ldr       w3, comps_cnt
+
+pcc_check_start:
+    ldr       c4, [x1, #COMP_OFFSET_PCC] // get current stored PCC to check
+    gcvalue   x4, c4
+    cmp       x4, x0 // check value of current PCC against given function
+    b.eq      pcc_found
+    add       w2, w2, #1
+    cmp       w2, w3 // compartment bounds check
+    b.eq      pcc_not_found
+    add       x1, x1, #COMP_SIZE // go check next PCC
+    b         pcc_check_start
+
+pcc_not_found:
+    b         abort
+
+pcc_found:
+    // We found the address to the start of the compartment info area, at
+    // address `$x1`
+    stp       x1, lr, [sp, #-16]! // store `x1` and `lr`
+    ldr       c2, [x1, #COMP_OFFSET_DDC]
+
+    // Unmap memory region allocated
+    gcvalue   x0, c2
+    gclen     x1, c2
+    bl        munmap
+    cmp       x0, #-1
+    b.eq      abort // munmap failed
+    ldp       x0, lr, [sp], #16
+
+    // Shift subsequent capabilities forward (if we delete the last
+    // compartment, we put gibberish instead)
+    ldr       w5, comps_cnt
+    mov       x6, #COMP_SIZE
+    ldr       x7, comps_addr
+    madd      x5, x5, x6, x7 // Get last allocated compartment address
+shift_next_comp:
+    add       x1, x0, #COMP_SIZE
+    ldp       c2, c3, [x1] // load next (DDC, PCC)
```

Comment:
I think you could just use `[x0, #COMP_SIZE]` and avoid maintaining `x1` entirely.

Reply:
I think this is not possible with the current logic due to line 188: `ldr x2, [x1]`. I'm essentially using `x1` as an internal pointer to the structure of the next compartment, the one we use to overwrite the current (deleted) compartment. So basically I need to load the value at `x0 + COMP_SIZE + COMP_OFFSET_STK_ADDR`, then store it to `x0 + COMP_OFFSET_STK_ADDR`, and I'm unsure how I can achieve this without using another register (mainly since needing to add `COMP_SIZE` and `COMP_OFFSET_STK_ADDR`, while maintaining `x0`).

Reply:
Couldn't you use `[x0, #(COMP_SIZE + COMP_OFFSET_STK_ADDR)]` there, in the same way?

```
    ldp c2, c3, [x0, #COMP_SIZE]
    stp c2, c3, [x0]
    ldr x2, [x0, #(COMP_SIZE + COMP_OFFSET_STK_ADDR)]
    str x2, [c0, #COMP_OFFSET_STK_ADDR]
    add ...

Reply:
If not, just disregard my comment. The code as it is appear correct.

Reply:
I didn't know you can write that syntax. Improved in https://github.com/capablevms/cheri-morello-compartmentalisation/pull/2/commits/fa69c9cf5f59744e792d30e0d7a5d00210b4d005.